# Improving Employee Retention by Predicting Employee Attrition Using Machine Learning

## Background

A technology startup is experiencing high employee turnover and the company lacks a deep understanding of the primary reasons why employees are choosing to resign. We assumed to be working in the Data Scientist team and we are expected to provide comprehensive data analysis to identify the contributing factors.

## Problem

The primary problem facing the company is high employee turnover. This not only increases recruitment and training costs but can also disrupt team productivity and ongoing projects. The company needs effective solutions to reduce turnover and improve employee retention.

## Goals

Identify the key factors tha lead to employees to resign and build a predictive model to anticipate employees who are likely to resign in the future. This model will enable the company to take proactive measures.

## Data

In [1]:
# import libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

In [2]:
df = pd.read_excel('Improving Employee Retention by Predicting Employee Attrition Using Machine Learning.xlsx')

# EDA  

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 25 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Username                            287 non-null    object 
 1   EnterpriseID                        287 non-null    int64  
 2   StatusPernikahan                    287 non-null    object 
 3   JenisKelamin                        287 non-null    object 
 4   StatusKepegawaian                   287 non-null    object 
 5   Pekerjaan                           287 non-null    object 
 6   JenjangKarir                        287 non-null    object 
 7   PerformancePegawai                  287 non-null    object 
 8   AsalDaerah                          287 non-null    object 
 9   HiringPlatform                      287 non-null    object 
 10  SkorSurveyEngagement                287 non-null    int64  
 11  SkorKepuasanPegawai                 282 non-n

The data has a total of 25 columns and 287 rows.
Columns 'Username', 'EnterpriseID', 'Email', 'NomorHP' are the columns unique to each employees. 
- 'Username': Username for each employee
- 'EnterpriseID': ID for each employee
- 'StatusPernikahan': Marriage status 
- 'JenisKelamin': Gender
- 'StatusKepegawaian': Employee status
- 'Pekerjaan': Job position
- 'JenjangKarir': Progression of an employee's career
- 'PerformancePegawai': Employee's performance
- 'AsalDaerah': Area of origin 
- 'HiringPlatform': The platform where employees got hired
- 'SkorSurveyEngagement': Engagement survey score
- 'SkorKepuasanPegawai': Employee satisfaction score
- 'JumlahKeikutsertaanProjek': Number of project participation
- 'JumlahKeterlambatanSebulanTerakhir': Number of Late Arrivals in the Past Month
- 'JumlahKetidakhadiran': Number of Absences
- 'NomorHP': Phone number
- 'Email': Email 
- 'TingkatPendidikan': Level of Education 
- 'PernahBekerja': Previously Worked
- 'IkutProgramLOP': Participated in the LOP Program
- 'AlasanResign': Reason for Resignation
- 'TanggalLahir': Date of Birth
- 'TanggalHiring': Date of Joining the company
- 'TanggalPenilaianKaryawan': Date of Employee Performance Review
- 'TanggalResign': Date of Resignation

In [4]:
df.sample(5)

Username  EnterpriseID StatusPernikahan JenisKelamin  \
70       blissfulGnu4        106501          Lainnya       Wanita   
138      thriftyMoth4        101511          Lainnya         Pria   
207   troubledQuiche5        106480          Menikah       Wanita   
139    gleefulKitten1        106347    Belum_menikah         Pria   
189  grudgingMeerkat3        106008    Belum_menikah       Wanita   

    StatusKepegawaian                      Pekerjaan           JenjangKarir  \
70          Outsource    Software Engineer (Android)              Mid_level   
138          FullTime   Software Engineer (Back End)  Freshgraduate_program   
207          FullTime  Software Engineer (Front End)              Mid_level   
139          FullTime   Software Engineer (Back End)  Freshgraduate_program   
189         Outsource   Software Engineer (Back End)  Freshgraduate_program   

    PerformancePegawai       AsalDaerah HiringPlatform  ...         NomorHP  \
70              Kurang  Jakarta Selatan         Indeed  ...  +6283206252xxx   
138             Kurang  Jakarta Selatan  Google_Search  ...  +6289570871xxx   
207       Sangat_bagus    Jakarta Pusat         Indeed  ...  +6285896075xxx   
139       Sangat_bagus    Jakarta Pusat  Google_Search  ...  +6287713961xxx   
189              Bagus    Jakarta Timur  CareerBuilder  ...  +6285714869xxx   

                              Email  TingkatPendidikan  PernahBekerja  \
70      blissfulGnu4786@hotmail.com           Magister              1   
138      thriftyMoth4270@icloud.com            Sarjana              1   
207   troubledQuiche5924@icloud.com            Sarjana              1   
139   gleefulKitten1142@outlook.com            Sarjana              1   
189  grudgingMeerkat3383@proton.com            Sarjana              1   

     IkutProgramLOP   AlasanResign TanggalLahir TanggalHiring  \
70              NaN  masih_bekerja   1986-07-24    2012-11-05   
138             NaN  masih_bekerja   1974-12-21    2009-01-05   
207             NaN     leadership   1972-11-21    2017-01-07   
139             NaN  masih_bekerja   1988-03-17    2011-01-10   
189             NaN            NaN   1986-04-17    2011-01-10   

    TanggalPenilaianKaryawan  TanggalResign  
70                 2020-2-13              -  
138               2019-02-05              -  
207               2020-02-06      2013-5-30  
139               2018-04-06              -  
189               2017-02-03              -  

[5 rows x 25 columns]

In [5]:
cat_var = df.drop(columns=['Username','EnterpriseID','NomorHP','Email'])


for i in cat_var:
    print("-" * 30) 
    print(f'Columns: {i}')
    print(cat_var[i].value_counts())
print("=" * 30) 

------------------------------
Columns: StatusPernikahan
StatusPernikahan
Belum_menikah    132
Menikah           57
Lainnya           48
Bercerai          47
-                  3
Name: count, dtype: int64
------------------------------
Columns: JenisKelamin
JenisKelamin
Wanita    167
Pria      120
Name: count, dtype: int64
------------------------------
Columns: StatusKepegawaian
StatusKepegawaian
FullTime      217
Outsource      66
Internship      4
Name: count, dtype: int64
------------------------------
Columns: Pekerjaan
Pekerjaan
Software Engineer (Back End)      109
Software Engineer (Front End)      72
Software Engineer (Android)        24
Product Design (UI & UX)           24
Product Manager                    17
Data Analyst                       16
Data Engineer                      10
Scrum Master                        3
Software Engineer (iOS)             3
DevOps Engineer                     3
Digital Product Manager             2
Machine Learning Engineer           2
Pro

In [6]:
# Change data types to data time

timestamp_var = ['TanggalLahir','TanggalHiring','TanggalPenilaianKaryawan']
for i in timestamp_var:
    df[i] = pd.to_datetime(df[i])
    

There is '-' value in 'TanggalResign', hence, it cant be replaced with date time data type

In [7]:
df.describe()

EnterpriseID  SkorSurveyEngagement  SkorKepuasanPegawai  \
count     287.000000            287.000000           282.000000   
mean   105923.324042              3.101045             3.904255   
min    100282.000000              1.000000             1.000000   
25%    101269.000000              3.000000             3.000000   
50%    106069.000000              3.000000             4.000000   
75%    110514.500000              4.000000             5.000000   
max    111703.000000              5.000000             5.000000   
std      4044.977599              0.836388             0.913355   

       JumlahKeikutsertaanProjek  JumlahKeterlambatanSebulanTerakhir  \
count                 284.000000                          286.000000   
mean                    1.179577                            0.412587   
min                     0.000000                            0.000000   
25%                     0.000000                            0.000000   
50%                     0.000000                            0.000000   
75%                     0.000000                            0.000000   
max                     8.000000                            6.000000   
std                     2.294441                            1.275016   

       JumlahKetidakhadiran  IkutProgramLOP                   TanggalLahir  \
count            281.000000       29.000000                            287   
mean              10.448399        0.517241  1979-02-14 01:15:15.679442496   
min                1.000000        0.000000            1951-01-02 00:00:00   
25%                5.000000        0.000000            1973-12-03 00:00:00   
50%               10.000000        1.000000            1980-09-30 00:00:00   
75%               15.000000        1.000000            1986-05-31 00:00:00   
max               55.000000        1.000000            1992-08-17 00:00:00   
std                6.902252        0.508548                            NaN   

                       TanggalHiring       TanggalPenilaianKaryawan  
count                            287                            287  
mean   2013-02-04 23:14:50.592334592  2018-09-29 18:48:55.191637504  
min              2006-01-09 00:00:00            2011-07-14 00:00:00  
25%              2011-07-11 00:00:00            2017-05-03 12:00:00  
50%              2013-04-01 00:00:00            2020-01-16 00:00:00  
75%              2014-08-18 00:00:00            2020-02-04 00:00:00  
max              2018-07-09 00:00:00            2020-02-28 00:00:00  
std                              NaN                            NaN

In [8]:
# Check Duplicate Value
df[df.duplicated()]

Empty DataFrame
Columns: [Username, EnterpriseID, StatusPernikahan, JenisKelamin, StatusKepegawaian, Pekerjaan, JenjangKarir, PerformancePegawai, AsalDaerah, HiringPlatform, SkorSurveyEngagement, SkorKepuasanPegawai, JumlahKeikutsertaanProjek, JumlahKeterlambatanSebulanTerakhir, JumlahKetidakhadiran, NomorHP, Email, TingkatPendidikan, PernahBekerja, IkutProgramLOP, AlasanResign, TanggalLahir, TanggalHiring, TanggalPenilaianKaryawan, TanggalResign]
Index: []

[0 rows x 25 columns]

There is no duplicate data in this dataset

In [9]:
# Check Missing Value
df.isnull().sum()

Username                                0
EnterpriseID                            0
StatusPernikahan                        0
JenisKelamin                            0
StatusKepegawaian                       0
Pekerjaan                               0
JenjangKarir                            0
PerformancePegawai                      0
AsalDaerah                              0
HiringPlatform                          0
SkorSurveyEngagement                    0
SkorKepuasanPegawai                     5
JumlahKeikutsertaanProjek               3
JumlahKeterlambatanSebulanTerakhir      1
JumlahKetidakhadiran                    6
NomorHP                                 0
Email                                   0
TingkatPendidikan                       0
PernahBekerja                           0
IkutProgramLOP                        258
AlasanResign                           66
TanggalLahir                            0
TanggalHiring                           0
TanggalPenilaianKaryawan          

Missing values can be seen on several columns; 
- 'SkorKepuasanPegawai' has minimal missing values (5) and the overall data distribution is near normal, hence, the missing values will be filled with median as it is more robust.
- 'JumlahKeikutsertaanProjek', 'JumlahKeterlambatanSebulanTerakhir' and 'JumlahKetidakhadiran' also have minimal missing values, (3), (1) and (6), respectively. Median will be used to fill in missing values.
- 'AlasanResign'= In this columns, there are 66 missing values, which may implies to the employees that are still working in the company. The missing value (66) added up with the 'masih_bekerja' value (132), equal to 198 which is in line with the column 'TanggalResign' that has '-' values. Hence, the missing value will be filled with 'masih_bekerja'
- 'IkutProgramLOP' has the most missing values, almost 90% of its data. Hence, the missing value will be replaced with unknown value. 
- In column 'StatusPernikahan', there is 3 '-' values which will be replaced with the column's mode. 

In [10]:
df['StatusPernikahan'] = df['StatusPernikahan'].replace('-',df['StatusPernikahan'].mode().iloc[0])
df['AlasanResign'] = df['AlasanResign'].fillna(df['AlasanResign'].mode().iloc[0])
df['SkorKepuasanPegawai'] = df['SkorKepuasanPegawai'].fillna(df['SkorKepuasanPegawai'].median())
df['JumlahKeikutsertaanProjek'] = df['JumlahKeikutsertaanProjek'].fillna(df['JumlahKeikutsertaanProjek'].median())
df['JumlahKeterlambatanSebulanTerakhir'] = df['JumlahKeterlambatanSebulanTerakhir'].fillna(df['JumlahKeterlambatanSebulanTerakhir'].median())
df['JumlahKetidakhadiran'] = df['JumlahKetidakhadiran'].fillna(df['JumlahKetidakhadiran'].median())
df['IkutProgramLOP'] = df['IkutProgramLOP'].fillna('Unknown')

In [11]:
df.PernahBekerja.value_counts()

PernahBekerja
1      286
yes      1
Name: count, dtype: int64

In [12]:
# drop column PernahBekerja as it has only one unique value
df_drop = df.drop(columns='PernahBekerja')